rust/agama-logs/src/main.rs

```diff
@@ -0,0 +1,215 @@
+extern crate fs_extra;
+extern crate tempdir;
+
+use fs_extra::copy_items;
+use fs_extra::dir::CopyOptions;
+use nix::unistd::Uid;
+use std::fs;
+use std::fs::File;
+use std::io;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use tempdir::TempDir;
+
+const DEFAULT_COMMANDS: [&str; 3] = [
+    "journalctl -u agama",
+    "journalctl -u agama-auto",
+    "journalctl --dmesg",
+];
+
+const DEFAULT_PATHS: [&str; 14] = [
+    // logs
+    "/var/log/YaST2",
+    "/var/log/zypper.log",
+    "/var/log/zypper/history*",
+    "/var/log/zypper/pk_backend_zypp",
+    "/var/log/pbl.log",
+    "/var/log/linuxrc.log",
+    "/var/log/wickedd.log",
+    "/var/log/NetworkManager",
+    "/var/log/messages",
+    "/var/log/boot.msg",
+    "/var/log/udev.log",
+    // config
+    "/etc/install.inf",
+    "/etc/os-release",
+    "/linuxrc.config",
+];
+
+const DEFAULT_RESULT: &str = "/tmp/agama_logs";
+const DEFAULT_NOISY: bool = true;
+const DEFAULT_COMPRESSION: (&str, &str) = ("bzip2", "tar.bz2");
+const DEFAULT_TMP_DIR: &str = "agama-logs";
+
+// A wrapper around println which shows (or not) output depending on noisy boolean variable
+macro_rules! showln
```

Comment:
eh ... why not? println which is wrapped inside is a macro too. So implementing a function which wraps a macro sounds ... weird to me.

Reply:
I do not think it is weird. The question is: why do you need a macro? Why a macro offers in this use case that a function does not?

Reply:
well i don't know exactly how it is in rust but i guess it is similar to C at least a bit.

So macros are replaced during compilation by its body. Functions are runtime thing which uses all that handling of params and returns using the stack and so on.  So in my POV macros are something "simplier" than functions and functional envelope around a macro is just adding unnecessary (internal) complexity.

Reply:
I'm afraid I have to disagree. In Rust, you should use a macro when a function falls short. Otherwise, you might write most of your code using macros, which would be hard to read and maintain.

IMHO the only justification to write a macro here is that you want to use a variable number of arguments. That's the reason for `println!` to be a macro: Rust does not support variadic functions. However, I would most probably use an already existing macro for that:

```rust
showln(true, format!("Storing result {}", true))
```

About being `simpler`, I find:

```rust
macro_rules! showln
{
    ($n:expr, $($arg:tt)*) => { if($n) { println!($($arg)*) } }
}
```

harder to read than:

```rust
fn showln(show: boolean, text: &str) {
  if (show) { println!(text) }
}
```

However, I am not going to block the PR because of this. But, please, document the macro.

Reply:
honestly, I don't care. One of my motivation was to try implement macro(s) in rust  as they are way too more complex then in C. So, I did and my soul is calm now. So, I can re-write it into function ... that should be simple enough ;-)